### ISSUE TITLE
About training pSp encoder results

### ISSUE THREAD

**Issue by pnbao** [eladrich/pixel2style2pixel#77]
Thank you for your hard work and enthusiastic support!!

I'm training the pSp encoder for encoding human face into baby face. The training logs look pretty well but I have noticed some weird results. 
If the face in original image is not facing the front, the model seems like trying to frontalize the final result.

<img width="795" alt="Screen Shot 2021-01-28 at 16 34 54" src="https://user-images.githubusercontent.com/11147999/106107145-e7a9a280-6189-11eb-80bc-52fab10bfe42.png">

<img width="788" alt="Screen Shot 2021-01-28 at 17 15 09" src="https://user-images.githubusercontent.com/11147999/106109070-70294280-618c-11eb-9d9b-179cb9826ec3.png">

I'm training the encoder with the below script and `paired dataset (original-faces, baby-faces)` on both train set and test set, `pretrained_models/psp_ffhq_encode.pt` is used as checkpoint.

```
python scripts/train.py \
--dataset_type=ffhq_encode
--exp_dir=/baby-experiment
--workers=8 
--batch_size=8 
--test_batch_size=8 
--test_workers=8 
--val_interval=2500 
--save_interval=5000 
--encoder_type=GradualStyleEncoder 
--start_from_latent_avg 
--lpips_lambda=0.8 
--l2_lambda=1 
--id_lambda=0.1 
--stylegan_weights=pretrained_models/baby_stylegan.pt 
--checkpoint_path=pretrained_models/psp_ffhq_encode.pt
```

The `pretrained_models/baby_stylegan.pt`  is baby faces generator which is trained with StyleGAN2.

One more thing is that all images' background in the dataset I'm using is removed as shown in the above image, but the encoded results always have gray background.
  
Do you have any thoughts about these problem? 
Thank you :)

**Comment by yuval-alaluf** [comment]
Hi @pnbao ,
We appreciate the positive feedback! 
Regarding your questions, it's interesting why you're seeing those artifacts on non-frontal images.
Now, while I am not sure if this is causing the problem, I have a feeling it has something to do with the `--dataset_type=ffhq_encode`. 
Notice that when defining the transforms for the dataset type we set the `transform_source` to None:
https://github.com/eladrich/pixel2style2pixel/blob/0730e33dc7f2aeafc67579c4e66dd7a5df0ceeb1/configs/transforms_config.py#L21-L38
Then, when we define the dataset we have: 
https://github.com/eladrich/pixel2style2pixel/blob/0730e33dc7f2aeafc67579c4e66dd7a5df0ceeb1/datasets/images_dataset.py#L28-L31
and we set `from_im = to_im`, which is not what you want in this case because instead of outputting the "adult" image, you're inputing the "baby" image into pSp. (Please verify me and let me know if I am wrong :smile:) 
I need to think a bit about why this would cause the weird artifact you're seeing in the first image, maybe this is the main problem. 

In any case, if this is indeed the problem, you can define your own dataset type and define the correct transforms in your case which would ensure that the correct input and target images are set.

**Comment by pnbao** [comment]
Thank you for quick reply!!

Actually I have modified the `transform_source` as your https://github.com/eladrich/pixel2style2pixel/issues/21#issuecomment-717403921, sorry I forgot that modification 😄


<img width="672" alt="Screen Shot 2021-01-28 at 18 11 48" src="https://user-images.githubusercontent.com/11147999/106116121-e467e400-6194-11eb-9d26-4903a1747084.png">

Was this caused the problem?

**Comment by yuval-alaluf** [comment]
I think so! 
Notice that in `transform_gt_train` you have a random horizontal flip but in the source transform you don't. Therefore, sometimes you're randomly flipping the target image without correspondingly flipping the source image.  
This may explain why you're seeing the artifact on non-frontal images more than the frontal images since this horizontal flip will affect non-frontal images much more than frontal ones. 
A quick fix would be to remove the RandomHorizontalFlip in the `transform_gt_transform`. Another option would be to ensure that if you're using the horizontal flip, then both the source and target are transformed together. 
I hope this makes sense :smile:

**Comment by pnbao** [comment]
Ah thank you I understand now 😄
I will try by removing the RandomHorizontalFlip in the `transform_gt_train`.
On the other hand, how to make sure `both the source and target are transformed together` as they are randomly horizontal flip?

**Comment by yuval-alaluf** [comment]
For transforming the source and target images together, you'll probably need to define your own transform class that receives both the images as inputs and randomly performs a flip on both of them together. This is not implemented in our code, but it should be easy to add this feature if you wish.

**Comment by pnbao** [comment]
I got better results when re-training without the RandomHorizontalFlip.

<img width="768" alt="Screen Shot 2021-01-29 at 10 56 19" src="https://user-images.githubusercontent.com/11147999/106221324-c647c580-6220-11eb-81e9-041f658c1aca.png">

I think it's meet the requirement now 😄 Thank you for your quick support!!
